pkg/admission/admission.go

```diff
@@ -0,0 +1,112 @@
+package admission
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// Config configures a Controller.
+type Config struct {
+	Limit uint64
+}
+
+// Default config settings:
+const (
+	defaultConcurrency = 50
+)
+
+// DefaultConfig returns a Config with preset values.
+func DefaultConfig() Config {
+	c := Config{
+		Limit: defaultConcurrency,
+	}
+
+	return c
+}
+
+// Controller determines the current admission parameters.
+type Controller struct {
+	Pool *quotapool.IntPool
+
+	mu struct {
+		syncutil.RWMutex
+
+		numAcquired int64
+		numWaiting  int64
+	}
+}
+
+// NewController constructs a new Controller struct from a given Config.
+func NewController(conf Config) *Controller {
+	fmt.Println("initializing new controller")
+	c := &Controller{
+
+		// Should this have a better name?
+		Pool: quotapool.NewIntPool("controller intpool", conf.Limit),
+	}
+
+	// Run some goroutine w/ stopper to log stats somewhere.
+	fmt.Printf("intpool cap: %d\n", c.Pool.Capacity())
+	return c
+}
+
+// NumWaiting returns the current number of requests that have been intercepted
+// by admission.Interceptor but have not been acquired.
+func (c *Controller) NumWaiting() int64 {
+	return atomic.LoadInt64(&c.mu.numWaiting)
+}
+
+// NumAcquired returns the number of requests we have acquired quota for.
+// This is functionally identical to computing
+// c.pool.Capacity() - c.pool.ApproximateQuota().
+func (c *Controller) NumAcquired() int64 {
+	return atomic.LoadInt64(&c.mu.numAcquired)
+}
+
+func (c *Controller) handleRelease(alloc *quotapool.IntAlloc) {
+	atomic.AddInt64(&c.mu.numAcquired, -1)
+	c.Pool.Release(alloc)
+}
+
+// Interceptor returns a UnaryServerInterceptor with parameters
+// taken from a Controller.
+func Interceptor(c *Controller) grpc.UnaryServerInterceptor {
+
+	admissionInterceptor := func(ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler) (interface{}, error) {
+
+		atomic.AddInt64(&c.mu.numWaiting, 1)
+
+		if info.FullMethod == "/cockroach.roachpb.Internal/Batch" {
+
```

Comment:
I suspect that there are certain requests which you're going to need to allow to bypass this altogether. My sense is that there are certain admin requests which are getting stuck here and thus causing the throughput to hit zero. Things like lease requests and node-liveness heartbeats, amongst others, really do need to get through. 

I think instead what you can do is type assert the argument and then verify some properties:

```
if ba, ok := req.(*roachpb.BatchRequest); ok { // instead of looking at the string

    // Allow requests that are certainly going to system ranges to succeed.
    // This is insufficient to determine whether a request is going to a system range but it's something and handles the most important cases.
    const maxCertainSystemRangeID = 20
    if ba.IsAdmin() || ba.Header.RangeID < maxCertainSystemRangeID /* total hack but curious to see what it does */ {
        return handler(req, info)
    }
    // ... logic you wrote

Reply:
Incorporated. I was not aware of Go's type-assertion syntax. Neat.

This seems like a reasonable idea. I'm not seeing many of these outages anymore, but I'll document those separately.